### DISCUSSION TITLE
How to preserve colorspace?

### DISCUSSION THREAD

**Discussion by gregbenz** [ImageMagick/ImageMagick#5445] [Digital Image Processing]
I'm doing some testing with a 32-bit TIF image with an embedded Rec2020 colorspace. When I convert my image using the commands below to create a JXL and then bring it back to TIF, the image has been significantly degraded. I'm doing a double conversion to help confirm that things are working properly, as well as to test some image formats which I cannot validate in Photoshop (ie, if it comes back the same then I know the export is good even if I cannot view it currently). 

```
magick "originalSource.tif" -depth 32 -define quantum:format=floating-point "converted.hdr"

magick "converted.hdr" -depth 32 -define quantum:format=floating-point "reversedToConfirm.tif"
```

What I see is that "converted.hdr" has changed from my Rec2020 embedded color space to an sRGB colorspace. It does show HDR data, but with substantial loss of data (very crude highlights and other tonal shifts). 

The "reversedToConfirm.tif" is further degraded and has no embedded profile at all.

What commands should be used to keep the exact same colorspace and pixels when converting from HDR to TIF and back? 

Note: I am using ImageMagick 7.1.0-45 Q16-HDRI x64 e32676e:20220731

**Comment by fmw42** [comment]
See A Word About Color Spaces at https://imagemagick.org/script/formats.php where it says that HDR only supports linear RGB. May not be relevant if your Rec2020 color space is linear RGB or it could be an oversight.  But you would be better assigning profiles for it rather than rely upon it being copied as such. 

See the list of colorspaces at https://imagemagick.org/script/command-line-options.php#colorspace. I do not see Rec2020. So it likely is not supported and is subsituted with RGB.  However, you could request an enhancement for it. But again a profile may be better.

The IM developers will need to comment further about this

**Reply by gregbenz** [reply]
Thank you, @fmw42

My source file is linear (Photoshop automatically uses linear gamma for all images converted to 32-bits and the profile is labeled that way). My source file was created by starting with a 16-bit image in Rec2020 with a gamma of 2.2 or 2.4 and converted to 32-bits. I don't know how to extract the embedded profile Photoshop creates, so possible it somehow mangles the gamma tag in the conversion to linear (but seems unlikely). Even if were to start directly in 32-bit, Photoshop asks for a colorspace and will linearize it. So I don't know a way to test this, but my assumption is that the embedded ICC profile for a 32-bit document saved by PS is written in a way which is standards compliant.

The list of colorspaces there reads to me more like color models/modes. More importantly, none of these strike me as being inherently wide color gamut spaces I could use (such as Rec2020, P3, ProPhoto, or Adobe RGB). I could use sRGB, but most HDR work is also going to venture into wide gamut, so that would be a problematic option.

**Reply by fmw42** [reply]
Please show what you get from 

magick identify -verbose image.tif 

Does it list the profile near the bottom of the listing? If there is an embedded profile, then 

magick image.tif icc:Rec2020.icc

should save it.

**Reply by gregbenz** [reply]
When I extract the ICC profile from my TIF using the command you suggested, the resulting profile (as reviewed in ColorSync utility) has a profile name matching PS exactly and the gamma is linear.

When I run ImageMagick's identify on my source TIF, I get the following (I've removed lines which seem irrelevant):
Several things seem wrong: 

- lists the colorspace as sRGB. Photoshop's info panel shows "REC2020 GAMMA 2.2 (linear RGB profile)"
- gamma is listed as 0.4545, which is 1/2.2. 
- says 16-bits per RGB channel for my 32 bit document

 ```
 Format: TIFF (Tagged Image File Format)
  Mime type: image/tiff
  Colorspace: sRGB
  Type: TrueColor
  Depth: 32/16-bit
  Channel depth:
    Red: 16-bit
    Green: 16-bit
    Blue: 16-bit
  Gamma: 0.454545
  Profiles:
    Profile-8bim: 5952 bytes
    Profile-icc: 516 bytes
    Profile-iptc: 125 bytes
    Profile-xmp: 17130 bytes
  Version: ImageMagick 7.1.0-45 Q16-HDRI x64 e32676e:20220731 https://imagemagick.org
```

So ImageMagick is pulling a correct ICC profile out of the image, but its "identify" command is returning bad data (does not match the extracted profile nor what PS says it is reading and writing).

**Reply by fmw42** [reply]
Why do you say it is wrong?  The colorspace is not the same as the profile. The profile is added to convert the colorspace to the profile. The name of the icc profile might show in the data. Did you not find it?

magick image.tif -format "%[profile:icc]" info:

should list the name.

**Reply by gregbenz** [reply]
It strikes me as "wrong" because:

- sRGB is a colorspace and I can't think of any reason it would be referenced if it was not interpreted to be the profile applied to the image. 
- After conversion, Photoshop is listing the profile as sRGB, which seems to confirm that ImageMagic is treating the colorspace as sRGB. 
- My converted image shows a significantly smaller gamut, which feels consistent with a conversion to sRGB, but I have not converted manually to compare.

In the full list of "identify" information was the following XMP data in my TIF (which I probably should have left in above, but is just XMP and not necessarily the embedded profile):
```
  Profile-xmp: 17130 bytes
  Properties:
    icc:copyright: Copyright 2022 Adobe Systems Incorporated
    icc:description: REC2020 GAMMA2.2 (Linear RGB Profile)
    photoshop:ColorMode: 3
    photoshop:ICCProfile: REC2020 GAMMA2.2 (Linear RGB Profile)
```

Running:
`magick "sourceFile.tif" -format "%[profile:icc]"`
results in an error:
`magick: MissingArgument `-format' at CLI arg 2 @ fatal/magick-cli.c/ProcessCommandOptions/678.`

**Reply by fmw42** [reply]
You left off "info:"

`magick "sourceFile.tif" -format "%[profile:icc]" info:
`

You need to re-assign the profile in your conversion, though IM may still be making a mistake on that end.

`magick image.tif -profile Rec2020.icc image.hdr
`
This assumes you had previously saved your profile from the tif image,.

However, the IM developers will need to review and comment further.

**Reply by fmw42** [reply]
What did you get for 

`magick identify -verbose image.hdr `

from Imagemagick conversion. Did it have the Rec2020 profile?  

I think you confuse colorspace with profile.  I think colorspace is like Photoshop ColorMode. 3 is RGB correct?

**Reply by gregbenz** [reply]
> You left off "info:"
> 
Thanks, when I do that I see it return the expected profile same as I see in PS:
`REC2020 GAMMA2.2 (Linear RGB Profile)`

**Reply by fmw42** [reply]
Just a shortcut for what you should get from 

`magick identify -verbose image`

in order to see what profile is imbedded.

**Reply by gregbenz** [reply]
> `magick image.tif -profile Rec2020.icc image.hdr ` This assumes you had previously saved your profile from the tif image,.
> 

When I use the following (with the good ICC profile I extracted), the result is an HDR file which opens showing as sRGB in Photoshop:
`magick "mySource.tif" -profile "Rec2020.icc" -depth 32 -define quantum:format=floating-point "convertedBack.hdr"`

If I put a typo in the ICC name, it throws an error. So it seems like it is trying to use the profile, but not producing the expected result.

**Reply by fmw42** [reply]
Is that sRGB the profile or the color mode in PS?  The color settings shows the default, not the actual imbedded profile, I believe. So where did you see sRGB.icc as the imbedded profile in PS?

See what EXIFTOOL says for the profile for you file?

**Reply by gregbenz** [reply]
> See what EXIFTOOL says for the profile for you file?

Shows as REC2020 GAMMA2.2 (Linear RGB Profile) from my 32-bit TIF, same as I see in PS

`
ExifTool Version Number         : 12.44
Bits Per Sample                 : 32 32 32
Photometric Interpretation      : RGB
Strip Offsets                   : (Binary data 159 bytes, use -b option to extract)
Samples Per Pixel               : 3
Sample Format                   : Float; Float; Float
Format                          : image/tiff
Color Mode                      : RGB
ICC Profile Name                : REC2020 GAMMA2.2 (Linear RGB Profile)
Photoshop Thumbnail             : (Binary data 3201 bytes, use -b option to extract)
Has Real Merged Data            : Yes
Writer Name                     : Adobe Photoshop
Reader Name                     : Adobe Photoshop 2022
Color Space                     : Uncalibrated
Profile CMM Type                : Adobe Systems Inc.
Profile Version                 : 2.4.0
Profile Class                   : Display Device Profile
Color Space Data                : RGB
Profile Connection Space        : XYZ
Profile Date Time               : 2022:08:12 16:03:46
Profile File Signature          : acsp
Primary Platform                : Apple Computer Inc.
CMM Flags                       : Not Embedded, Independent
Rendering Intent                : Perceptual
Connection Space Illuminant     : 0.9642 1 0.82491
Profile Creator                 : Adobe Systems Inc.
Profile ID                      : 0
Profile Copyright               : Copyright 2022 Adobe Systems Incorporated
Profile Description             : REC2020 GAMMA2.2 (Linear RGB Profile)
Red Matrix Column               : 0.67348 0.27904 -0.00194
Green Matrix Column             : 0.16568 0.67535 0.02998
Blue Matrix Column              : 0.12505 0.04561 0.79686
Media White Point               : 0.95047 1 1.08905
Red Tone Reproduction Curve     : (Binary data 14 bytes, use -b option to extract)
Green Tone Reproduction Curve   : (Binary data 14 bytes, use -b option to extract)
Blue Tone Reproduction Curve    : (Binary data 14 bytes, use -b option to extract)`

**Reply by gregbenz** [reply]
> Is that sRGB the profile or the color mode in PS? The color settings shows the default, not the actual imbedded profile, I believe. So where did you see sRGB.icc as the imbedded profile in PS?

When I mentioned sRGB, I meant ImageMagick reporting "Colorspace: sRGB". 

In Photoshop, sRGB is just a profile. If you use the default PS profile, it means sRGB IEC61966-2.1, but there are of source other versions of sRGB. But any way you look at it, it would be a specific color model without ambiguity when the profile is embedded, which is the case with my test image.

In Photoshop, when you refer to the color model, that's part of the image "mode". Any RGB profile would use the image mode. But a specific color profile (whichever flavor of sRGB this is) would include details on the primaries, gamma, etc that are more detailed than the mode. sRGB would never be a reference to the generic mode (Rec2020 is used in the RGB mode but it is not related to the sRGB profile)

So "Colorspace: sRGB" jumps out at me because it's definitely a reference to a color space (albeit ambiguous in the details of which sRGB if you really dig into it) and is definitely not the color space of the source image (which was the linearized Rec2020 that Photoshop created automatically from my Rec2020 gamma2.2 color space, which came from an ICC profile from Eizo).

**Comment by snibgo** [comment]
Regarding colorspaces and profiles:

IM has those two methods for defining in metdata what pixels values mean, and converting pixel values from one meaning to another.

Two systems is not better than one, but they came about in the long evolution of IM. We are where we are.

Most "colorspaces" are really colour *models*, such as XYZ, YIQ, HCL and so on. But some such as Adobe98, RGB, sRGB are *real* colorspaces, in the sense that they have colorimeric definitions.

IM does some limited automatic conversion between colorspaces, for example, IM knows that CMYK data can't be stored in PNG files, so will automatically convert it to sRGB first. But most operations are purely numeric. For example, we might subract an sRGB image from a HCL image, to get a (probably) meaningless result.

Near the top of verbose info we have:
```
Colorspace: sRGB
```
That refers to the "-colorspace" setting. It has no connection to any embedded profile.

IM has no "colorspace unknown" setting. Unless it knows otherwise, it assumes the colorspace is sRGB. For example, we might convert to a XYZ profile, and the "colorspace" setting will still claim the image is sRGB. If we want, we might "-set colorspace XYZ".

**Reply by gregbenz** [reply]
Any idea what would need to be done for the command above to convert from the 32-bit TIF to an HDR file which does not alter the embedded color space or the actual pixels? The shift I see is substantial.

**Comment by fmw42** [comment]
Did you try to add the profile as I suggested above.?

**Reply by gregbenz** [reply]
Yes, our exchange got a bit complicated. My reply above is https://github.com/ImageMagick/ImageMagick/discussions/5445#discussioncomment-3402266

The extracted profile is good, running the command executes without out, but the resulting HDR file does not have the requested profile. 

I just tried again and this command generates an HDR without using the requested external ICC profile I had extracted:
`magick "source.tif" -profile "Rec2020.icc" "converted.hdr"`

**Reply by fmw42** [reply]
But do you not have the previously extracted profile on your system in the working directory? If so, it will be used. If not, the output may be generated but without the profile.  Did you check the HDR to see if it had that profile embedded?

**Reply by gregbenz** [reply]
I'm using the correct path. When I try to use a bad path, I can confirm via an error message such as:
`magick: unable to read blob 'Rec2020-BAD.icc': No such file or directory @ error/blob.c/FileToBlob/1442.`

When I use the correct path and generate an HDR, inspecting that file with Magick lists not profile in the HDR
`magick identify -verbose "embeddedVersion.hdr"`

However, it now appears to me that the HDR format does not support an embedded profile (must use some standard). If I change the profile in PS and save as an HDR, that new file from PS won't show a profile through Magick and will reopen in PS with the generic sRGB profile listed - not the one I converted to in PS.

**Reply by gregbenz** [reply]
ImageMagick should probably show a warning or error when requesting a profile for an HDR file, to indicate that it is not a valid option.

**Reply by gregbenz** [reply]
I'm trying to find some format I can send to and back in order to confirm preservation of the data, but having no luck. If I send to JXL and back, the result shows significant loss of color (looks like clipping to sRGB gamut). The over-range HDR values are close enough but also changed during the round trip to JXL and back to 32-bit TIF.

If I try to send to PSD or PSB (which are nice because they perfectly preserve the data in Photoshop, including the profile), ImageMagick is converting my 32-bit source to 16-bits when using the following commands:
`magick "source.tif" -depth 32 -profile "Rec2020.icc" "newFile.psd"`
`magick "source.tif" -depth 32 -profile "Rec2020.icc" "newFile.psb"`

**Comment by snibgo** [comment]
You say you are using "7.1.0-45 Q16-HDRI x64".

>  ImageMagick is converting my 32-bit source to 16-bits when using the following commands: ...

If your source is 32-bit integer, then IM will rescale the values in memory to 16-bit integer.

> ImageMagick should probably show a warning or error when requesting a profile for an HDR file, to indicate that it is not a valid option.

IM doesn't request a profile for any particular file format. Perhaps what you mean is "When writing an image with embedded profile to HDR format, IM should warn that the profile won't be written."

Well, perhaps, but most file formats do not record all the metadata that IM has about an image. If IM warned about metadata that would *not* be written, IM would become very complex, and people would complain about the large number of messages.

**Reply by gregbenz** [reply]
Shouldn't it preserve the 32-bit data or warn the user? It isn't something which preserves image quality like a conversion from 32-bit integer to 16-bit floating point. The resulting 16-bit PSB/PSD file is significantly degraded.

Good points on complexity for the interface, I wouldn't expect warnings for many conversion scenarios. However in this case it is an explicit request which is being ignored. Shouldn't there be a warning when the user is requested a feature flag (or combination of flags) which is invalid?

**Reply by fmw42** [reply]
can you post a small example 32-bit tiff that you want to convert. Photoshop may rescale the values to the range 0 to 1, which when brought back to IM may still be in that range rather than 0 to 32-bit range. A rescale may be needed.

**Reply by gregbenz** [reply]
When I run the instructions below starting from the attached source (smallHdrTest.tif), the round trip files are significantly different.
- smallHdrTest-backFrom**JXL**.tif shows HDR values, but they are significantly less bright. And the color gamut has been significantly reduced.
- smallHdrTest-backFrom**AVIF**.tifis significantly clipped. It is not scaled and the data cannot be recovered, so this is substantially more degraded than the JXL pathway.


`magick "smallHdrTest.tif" -depth 32 -define quantum:format=floating-point "smallHdrTest.jxl"`

`magick "smallHdrTest.jxl" -depth 32 -define quantum:format=floating-point "smallHdrTest-backFromJXL.tif"`

`magick "smallHdrTest.tif" -depth 32 -define quantum:format=floating-point "smallHdrTest.avif"`

`magick "smallHdrTest.avif" -depth 32 -define quantum:format=floating-point "smallHdrTest-backFromAVIF.tif"`

Source image (and my conversions): https://drive.google.com/drive/folders/1DPqOc6nviGzOq-39T0lx2LvgRDWuGGTy?usp=sharing

If you open the TIF files in PS and use the 32-bit color picker, you can see the data clearly. To visualize the HDR data on any monitor, try the 32-bit exposure slider or add an exposure adjustment layer with -3 exposure. Or if you have a true HDR monitor, MacOS, and the latest PS (v23.4.2), go to PS prefs / tech previews, enable the HDR option and restart PS and then open the image.

**Reply by fmw42** [reply]
Do JXL and AVIF support 32-bits and moreover does Imagemagick support that in those formats? I do not know. One of the IM developers will need to answer.  Have you checked those format's documentation?

**Reply by gregbenz** [reply]
The JXL and AVIF format specifications support HDR content (though I don't recall if they are encoded at 32-bits or use a different encoding at a lower bit-depth to approximate the same data). Not clear on ImageMagick's support, that's what I'm trying to learn here.

**Comment by snibgo** [comment]
> However in this case it is an explicit request which is being ignored.

What explicit request is being ignored?

In your command ...
```
magick "source.tif" -profile "Rec2020.icc" "converted.hdr"
```
... the input file is read into memory. Then if the image has an embedded profile it is converted to a new profile, otherwise the profile is assigned to the image. Then the image is written to a file. These operations all succeed. No request is ignored.

`magick -list format` says:
```
      HDR* rw+   Radiance RGBE image format
```
Wikipedia suggests the defining document is https://floyd.lbl.gov/radiance/refer/filefmts.pdf , which describes the "Picture File Format", which has no facility to record ICC colour profiles. Each pixel is recorded with a total of 4 bytes (32 bits). It says:

> The accuracy of these values will be on the order of 1% (+/-1 in 200) over a dynamic range from 10^-38 to 10^38.

I would regard an accuracy of 1% as totally horrid. It is similar to 8-bit integer. Okay, HDR format has a wider dynamic range, but at great cost.

**Reply by gregbenz** [reply]
> What explicit request is being ignored?
I mean that there is an explicit request for a profile and the HDR file format, which does not support such a profile. It's not a big deal for me since I understand it, but it just seems like there should be feedback to the user when the requested settings are not valid.
`-profile "Rec2020.icc"`

My ultimate goal here is to get a conversion from a 32-bit TIF to an HDR image which can be viewed in a web browser. I don't plan to use the HDR format but had tried it merely as a way to test conversions I could review in Photoshop to see if I was at least getting to and from another 32-bit format without loss of data (since I'm new to ImageMagick, not confident with it, and have seen loss of quality with all my tests so far using JXL or AVIF). Given what I've learned and limits around color, HDR is not a file format I need to explore further.

What I'd really like to do is **generate valid JXL or AVIF files with HDR content**, as those are compact file formats which support HDR in theory. Browser support for HDR content isn't something I've found documented anywhere. I know that FireFox and Chrome support HDR video and they support these file formats, but I do not know if they support HDR versions of these image formats. Chrome support appears to be ahead of FireFox. So far I'm not seeing any HDR values in Chrome when viewing JXL or AVIF files in Chrome. My hypothesis is that Chrome is not supporting HDR content for JXL as I seem to be getting to and from JXL with HDR data in the round-tripped TIF. I don't know on AVIF as it does not look like I am getting HDR in the AVIF based on the round-tripped TIF - but I have been told that Chrome will support HDR content from AVIF (I just haven't found a test image or been able to create my own to confirm - let alone find a way to convert images to a format if it is supported in Chrome).

**Comment by snibgo** [comment]
> I mean that there is an explicit request for a profile and the HDR file format, which does not support such a profile. 

Okay, I see what you mean.

IM performs each operation separately. An image can be assigned or converted to an ICC profile, and the profile is embedded. An image (with or without an embedded ICC profile) can be saved to a file. If the file format doesn't support ICC profiles, then any embedded profiles won't be saved in the file. The same applies to other metadata; not all file formats can record all metadata.

These are separate operations. When an image with embedded profile is saved to a file, IM doesn't know or care if the profile was embedded earlier in the same command, or in a previous IM command, or by a camera, or whatever.

**Comment by gregbenz** [comment]
@snibgo Thank you for that explanation, makes sense.

Ultimately, I think the ideal solution for HDR processing will be to have support for JPG / AVIF gain maps with standard profiles (which in the case of an AVIF HDR base image includes PQ / HLG EOTF with common primaries, probably encoded as CICP rather than an ICC profile). At this point, those gain map formats are much more relevant than 32-bit TIF for tools intended primarily to support online sharing of images (I would not use 32-bit input or output with ImageMagick now, re-processing of an existing gain map is more relevant now to provide much better adaptation to displays of different capabilities).

Request for gain map support: https://github.com/ImageMagick/ImageMagick/issues/6377

**Reply by fmw42** [reply]
@gregbenz Pardon my ignorance, but why is an ICC profile not adequate and what are PQ / HLG EOTF these files and how are they used.

**Reply by gregbenz** [reply]
@fmw42 32-bit implies HDR (with linear encoding). There is currently no ICC standard for HDR and to the extent you can work around that in ICC (or embed CICP into an ICC profile), you'd have a niche profile that won't be well supported by browsers.

PQ and HLG are common EOTFs designed for HDR (you'd typically think of "gamma" for SDR images). This is part of a profile to describe how an image is encoded (a complete profile also needs to specify things like the color primaries, ie P3, sRGB, etc).

**Reply by fmw42** [reply]
Thanks

**Reply by blue-j** [reply]
FWIW, the ICC has had an active working group looking into HDR for quite some time. Let's hope for movement soon.
